service/kinesis/timeout_customization.go

```diff
@@ -0,0 +1,83 @@
+package kinesis
+
+import (
+	"io"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+)
+
+// readDuration is how long we will wait to read a block
+var readDuration = 2 * time.Minute
+
+var timeoutErr = awserr.New(
+	"RequestTimeout",
+	"read on body has reached the timeout limit",
+	nil,
+)
+
+type readResult struct {
+	n   int
+	err error
+}
+
+// timeoutReadCloser will handle body reads that take too long.
+// We will return a ErrReadTimeout error if a timeout occurs.
+type timeoutReadCloser struct {
+	reader   io.ReadCloser
+	duration time.Duration
+	retry    **bool
+}
+
+// Read will spin off a goroutine to call the reader's Read method. In addition,
+// we use the time.AfterFunc to return an timeout error once a duration of time
+// has occurred.
+func (r *timeoutReadCloser) Read(b []byte) (int, error) {
```

Comment:
Suggest doing a select instead of wait after

```go
func (r *timeoutReadCloser) Read(b []byte) (int, error) {
	t := time.NewTimer(dur)
	defer t.Stop()

	c := make(chan readResult)
	go func() {
		n, err := r.reader.Read(b)
		c <- readResult{n, err}
	}()

	select {
	case <-t.C:
		return 0, timeoutErr
	case res := <-c:
		return res.n, res.err
	}
}
```